Only enabled regions can interact with referrals journey

patinipat opened the pull request:
# Context

<!-- Is there a Trello ticket you can link to? -->
<!-- Do you need to add any environment variables? -->
<!-- Is an ADR required? An ADR should be added if this PR introduces a change to the architecture. -->

# Changes in this PR

* The referral journey is protected with by user.role and user.region
  Only referrer with `referrer` role and within an enabled region can create a referral

* Assessor will see the review referral tile when the have the correct role ``and are in an enabled region



## Screenshots of UI changes

### Before

### After
New static page /use-ndelius

![image](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/141184242/983ff1d4-801a-4a09-8474-a84bec80f9c4)


# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

patinipat committed:
Add user utils functions

* `isApplyEnabledForUser` returns true when user belongs to an enabled
region

* `userHasReferrerRole`, returns true when user is a referrer

* `userHasAssessorRole`, returns true when user is a assessor

* `userHasReferrerRoleAndIsApplyEnabled`, returns true when user is a
referrer and is in an apply enabeld region

* `userHasAssessorRoleAndIsApplyEnabled`, returns true when user is a
assessor and is in an apply enabeld region

patinipat committed:
Use `userHasAssessorRoleAndIsApplyEnabled` instead of feature flag `applyDisabled`

Assessors will only be able to see referrals when they are in an
enabled region.

Remove `applyDisabled` from nunjuck globals

patinipat committed:
Remove feature flags `applyDisabled` for applyRoutes

patinipat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Only expected regions can interact with referrals](https://trello.com/c/PBk6dQWJ/1361-only-expected-regions-can-interact-with-referrals)

bryangaledxw commented on server/utils/userUtils.test.ts:
```diff
@@ -22,3 +25,52 @@ describe('filterProbationRegions', () => {
+    [true, userFactory.build({ roles: ['assessor'], region: { name: 'Kent, Surrey & Sussex' } }), true],
+    [true, userFactory.build({ roles: [], region: { name: 'Kent, Surrey & Sussex' } }), false],
+    [true, userFactory.build({ roles: ['assessor'], region: { name: 'London' } }), true],
+    [true, userFactory.build({ roles: [], region: { name: 'London' } }), false],
```
Since these are all true, can it come out of the test parameters, and into the test itself?

And I feel like it'd be more readable if the parameters were the expected true/false, the roles array, the region, and you build the user inside the test body.

Then you can write the test name as something like `'returns %p when the roles are %p and the region is %p'`

patinipat replied:
:recycle:

bryangaledxw commented on server/utils/userUtils.ts:
```diff
@@ -1,6 +1,16 @@
+  return config.flags.allRegionsEnabled || !disabledRegions.includes(user.region.name)
+}
+
+export function canAssessorViewReferrals(user: User): boolean {
```
If feel like this should be `canUserViewReferrals`, as whether or not the user is an assessor is part of what this is testing

patinipat replied:
:recycle:

bryangaledxw commented on server/config.ts:
```diff
@@ -52,6 +52,7 @@ export default {
   flags: {
     oasysDisabled: process.env.OASYS_DISABLED || false,
     applyDisabled: !['local', 'dev', 'test', 'preprod'].includes(environment),
+    allRegionsEnabled: ['local', 'dev', 'test'].includes(environment),
```
Can this be more descriptive? `applyEnabledForAllRegions`?

patinipat replied:
:recycle:

bryangaledxw commented on server/middleware/roleMiddleware.ts:
```diff
@@ -1,14 +1,15 @@
 const wrapHandler =
   (handler: RequestHandler, role: UserRole) => async (req: Request, res: Response, next: NextFunction) => {
-    if (res.locals.user.roles.includes(role)) {
+    if (res.locals.user.roles.includes(role) && hasRegionEnabled(res.locals.user)) {
```
I think this needs a a rethink, as AFAICT this will end up restricting the entire app to users in our enabled regions. 

Tempted to say that the signature for `wrapHandler` should be:
`const wrapHandler = (handler: RequestHandler, check: (user: User) => bool))`...

Then instead of doing:
`if (res.locals.user.roles.includes(role))`
It does:
`if (check(res.locals.user)`

And createRoleMiddleware passes in a function, but even then I think you'd need to split out our routes for this to not break things for someone.

It may be we just don't do enforcement at this level

patinipat replied:
Added a new middleware createUserCheckMiddleware that is defined as suggested

bryangaledxw commented on server/@types/shared/models/TemporaryAccommodationUserRole.ts:
```diff
@@ -3,4 +3,4 @@
 /* eslint-disable */
 
-export type TemporaryAccommodationUserRole = 'assessor' | 'referrer';
+export type TemporaryAccommodationUserRole = 'assessor' | 'referrer' | 'cas3_referrer';
```
For my understanding, where's this change come from?

patinipat replied:
removed this commit based on our discussion

bryangaledxw commented on server/utils/disabledRegions.ts:
```diff
@@ -0,0 +1,13 @@
+export default [
```
My instinct is it'd be safer to have an enabled list rather than a disabled list. I kinda feel like doing this without an API change to get this info is unsafe in any case, but an enabled list feels less unsafe 🙂

patinipat replied:
:recycle:  
switched to an enabledRegions list

bryangaledxw commented on server/middleware/roleMiddleware.test.ts:
```diff
@@ -4,26 +4,57 @@ import { userFactory } from '../testutils/factories'
     const next = jest.fn()
+    config.flags.allRegionsEnabled = false
 
-    const middleware = createRoleMiddleware('referrer')
```
I feel like the change to `cas3_referrer` is worth its own commit

patinipat replied:
remove cas3_referrer based on the clarification that the backend api sends referrer and not cas3_referrer

bryangaledxw commented on server/controllers/temporary-accommodation/manage/dashboardController.test.ts:
```diff
@@ -1,11 +1,20 @@
 import type { NextFunction, Request, Response } from 'express'
```
In the message for this commit, "a referrals", and probably better to use "they" rather than "he"

patinipat replied:
:recycle:

bryangaledxw commented on server/utils/userUtils.ts:
```diff
@@ -1,6 +1,16 @@
   return [regions.find(region => region.id === request.session.probationRegion.id)]
 }
+
+export function hasRegionEnabled(user: User): boolean {
```
Similar comment to above on naming. Maybe `isApplyEnabledForUser`?

patinipat replied:
:recycle:

bryangaledxw reviewed (commented):
Good stuff on getting your head round this. I've left a fair few comments, but I think the only fundamental thing is the change in roleMiddleware, which I think would end up breaking things for existing users.

It's the routes under `routes/apply`, and some subset (the routes to AssessmentsController) of the routes under `routes/temparary-accommodation/manage` that we'd want to be limited by region. It may be we need to split `routes/temparary-accommodation/manage` into `manage` and `place`, or we simply don't enforce this for now at the routing level

bryangaledxw commented:
Just realised we also have LandingController, which will need to be updated

patinipat committed:
Add static page useNDelius

When referrers cannot access the apply journey they are redirected to
this static page

patinipat committed:
Add `userCheckMiddleware`

Accepts a user predicate function
When predicate succeeds the handler is called
When predicate fails returns the UnauthorizedError

This new middleware aims to replace `roleMiddleware`

patinipat committed:
Switch `createRoleMiddleware` for `createUserCheckMiddleware`

* No functional difference for manage routes

* No functional difference for apply routes

patinipat committed:
Update LandingController for Apply enabled region

when referrers from a disabled region attempts to access the
appliaciton journey they are redirected to the static page `use-ndelius`

bryangaledxw commented on server/views/temporary-accommodation/static/useNDelius.njk:
```diff
@@ -0,0 +1,19 @@
+{% extends "partials/layout.njk" %}
+
+{% set pageTitle = "Use nDelius" %}
```
Suggest this should be:
`{% set pageTitle = applicationName + " - Referrals should be submitted through nDelius" %}`
to match conventions elsewhere

patinipat replied:
:recycle:

bryangaledxw commented on server/views/temporary-accommodation/static/useNDelius.njk:
```diff
@@ -0,0 +1,19 @@
+{% set pageTitle = "Use nDelius" %}
+
+{% block content %}
+  <div class="govuk-grid-row">
```
I think in the GDS system you want rows, columns, then content, so this should contain something like:
`<div class="govuk-grid-column-two-thirds">`

patinipat replied:
used `<div class="govuk-grid-column-full">`

bryangaledxw commented on server/utils/userUtils.ts:
```diff
@@ -1,6 +1,28 @@
+  return config.flags.applyEnabledForAllRegions || enabledRegions.includes(user.region.name)
+}
+
+export function canUserViewReferrals(user: User): boolean {
```
Sorry, gonna be annoying about naming 😅 - think change to `userHasAssessorRoleAndIsApplyEnabled`, so we're staying consistent within `userUtils`

patinipat replied:
:recycle:

bryangaledxw commented on server/middleware/userCheckMiddleware.test.ts:
```diff
@@ -0,0 +1,48 @@
+    expect(handler).toHaveBeenCalledWith(request, response, next)
+  })
+
+  it('returns middleware that calls next() when the user predicate fails and no redirectPath provided', async () => {
```
Think the reference to `redirectPath` is a leftover

patinipat replied:
good spot

bryangaledxw commented on server/routes/apply.ts:
```diff
@@ -3,14 +3,19 @@
 import type { Controllers } from '../controllers'
 import Apply from '../form-pages/apply'
-import { createRoleMiddleware } from '../middleware/roleMiddleware'
+import { createUserCheckMiddleware } from '../middleware/userCheckMiddleware'
```
I think go ahead and delete roleMiddleware at this step

patinipat replied:
:+1:

bryangaledxw commented on server/controllers/temporary-accommodation/manage/dashboardController.ts:
```diff
@@ -1,10 +1,10 @@
 import type { Request, RequestHandler, Response } from 'express'
-import { canUserViewReferrals } from '../../../utils/userUtils'
+import { userHasAssessorRoleAndIsApplyEnabled } from '../../../utils/userUtils'
```
I think just have it as `userHasAssessorRoleAndIsApplyEnabled` from the first commit, rather than add the function and rename it later. dxw conventions are to structure your commits as though you knew the end result from the start, so it's telling an idealised story of how a feature was built. Definitely takes a bit of getting used to, but it keeps commit history clean and it makes a lot of difference when it comes to reviewing PRs

It might be worth dropping this commit entirely, and rolling its changes into the first two commits

patinipat replied:
:recycle:

bryangaledxw commented on server/routes/apply.ts:
```diff
@@ -5,7 +5,6 @@ import type { Controllers } from '../controllers'
 import Apply from '../form-pages/apply'
 import { createUserCheckMiddleware } from '../middleware/userCheckMiddleware'
 import paths from '../paths/apply'
-import staticPaths from '../paths/temporary-accommodation/static'
```
Think this belongs in the previous commit

patinipat replied:
:recycle:

bryangaledxw reviewed (commented):
Looks good. In terms of code changes it's pretty there. Think it's worth doing a sweep over the commit messages - I spotted "will protecting" in the first commit - and the commit tidying I've suggested. I'll give this some manual testing, as this is a risky one, but if you can do these last few changes, and fixing that test failure doesn't need any big changes, this should be good to go in.

bryangaledxw commented on server/routes/temporary-accommodation/index.ts:
```diff
@@ -22,10 +21,7 @@ export default function routes(controllers: Controllers, services: Services): Ro
 
   manageRoutes(controllers, services, router)
-
-  if (!config.flags.applyDisabled) {
```
Think remove this flag from `config.ts` entirely at this point, since it's no longer used

patinipat replied:
:recycle:

patinipat committed:
Update Manage Booking controller

Use the user helper function defined previously instead of the applyDisabled

bryangaledxw commented on server/controllers/temporary-accommodation/manage/bookingsController.test.ts:
```diff
@@ -30,12 +30,21 @@ import {
   insertBespokeError,
   insertGenericError,
 } from '../../../utils/validation'
+import { isApplyEnabledForUser } from '../../../utils/userUtils'
```
Think swap this commit with the previous one, so we're removing the dependence on `applyDisabled` being in `nunjucksSetup` before we remove `applyDisabled` from `nunjucksSetup`

patinipat replied:
:recycle:

bryangaledxw commented on server/controllers/temporary-accommodation/manage/bookingsController.test.ts:
```diff
@@ -30,12 +29,21 @@ import {
+  return {
+    ...module,
+    isApplyEnabledForUser: jest.fn(),
+  }
```
I wouldn't worry about this tbh, think just jest.mock('../../../utils/userUtils') is fine

There's a few odd cases in the codebase where we need a mocked version of a function and a real version in the same test file, but those are the only places we really use this pattern

patinipat replied:
:recycle:

bryangaledxw reviewed (approved):
Couple of last bits of feedback, but otherwise good for this to go in

On the merge commit at the end - we generally rebase on top of main, rather than doing merge commits, to keep the tree linear. Not sure if that'll make any difference once it's merged in though
